File: src/detection/lumaMetrics.ts
import type { NativeFrameAnalysisResult } from './nativeHeuristicTypes';

const TARGET_MEAN_LUMA = 0.52;
const MEAN_WEIGHT = 180;
const HIGHLIGHT_THRESHOLD = 0.96;
const SHADOW_THRESHOLD = 0.04;
const HIGHLIGHT_PENALTY = 80;
const SHADOW_PENALTY = 70;
const SHARPNESS_SCALE = 420;

export type LumaGrid = {
  width: number;
  height: number;
  values: readonly number[];
};

export type LumaMetricOptions = {
  createdAtMs?: number;
  targetMeanLuma?: number;
  sharpnessScale?: number;
};

export type LumaQualityMetrics = Pick<NativeFrameAnalysisResult, 'exposure' | 'sharpness'>;

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

function toUnitLuma(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return clamp(value > 1 ? value / 255 : value, 0, 1);
}

function assertGrid(grid: LumaGrid): void {
  if (!Number.isInteger(grid.width) || grid.width <= 0) throw new Error('Luma grid width must be positive.');
  if (!Number.isInteger(grid.height) || grid.height <= 0) throw new Error('Luma grid height must be positive.');
  if (grid.values.length !== grid.width * grid.height) throw new Error('Luma grid value count does not match width and height.');
}

export function computeLumaQualityMetrics(grid: LumaGrid, options: LumaMetricOptions = {}): LumaQualityMetrics {
  assertGrid(grid);

  let sum = 0;
  let highlightCount = 0;
  let shadowCount = 0;

  for (const raw of grid.values) {
    const value = toUnitLuma(raw);
    sum += value;
    if (value >= HIGHLIGHT_THRESHOLD) highlightCount += 1;
    if (value <= SHADOW_THRESHOLD) shadowCount += 1;
  }

  const sampleCount = grid.values.length;
  const meanLuma = sum / sampleCount;
  const clippedHighlightsRatio = highlightCount / sampleCount;
  const crushedShadowsRatio = shadowCount / sampleCount;
  const targetMeanLuma = options.targetMeanLuma ?? TARGET_MEAN_LUMA;
  const baseExposureScore = 100 - Math.abs(meanLuma - targetMeanLuma) * MEAN_WEIGHT;
  const exposureScore = clamp(baseExposureScore - clippedHighlightsRatio * HIGHLIGHT_PENALTY - crushedShadowsRatio * SHADOW_PENALTY, 0, 100);

  let edgeEnergyTotal = 0;
  let edgeSampleCount = 0;

  for (let y = 1; y < grid.height - 1; y += 1) {
    for (let x = 1; x < grid.width - 1; x += 1) {
      const center = y * grid.width + x;
      const left = toUnitLuma(grid.values[center - 1]);
      const right = toUnitLuma(grid.values[center + 1]);
      const up = toUnitLuma(grid.values[center - grid.width]);
      const down = toUnitLuma(grid.values[center + grid.width]);
      edgeEnergyTotal += Math.abs(right - left) + Math.abs(down - up);
      edgeSampleCount += 1;
    }
  }

  const edgeEnergy = edgeSampleCount === 0 ? 0 : edgeEnergyTotal / edgeSampleCount;
  const sharpnessScore = clamp(edgeEnergy * (options.sharpnessScale ?? SHARPNESS_SCALE), 0, 100);

  return {
    exposure: {
      exposureScore,
      meanLuma,
      clippedHighlightsRatio,
      crushedShadowsRatio
    },
    sharpness: {
      sharpnessScore,
      edgeEnergy
    }
  };
}

export function buildLumaQualityAnalysis(grid: LumaGrid, options: LumaMetricOptions = {}): NativeFrameAnalysisResult {
  const quality = computeLumaQualityMetrics(grid, options);

  return {
    status: 'low-confidence',
    createdAtMs: options.createdAtMs ?? Date.now(),
    subject: null,
    exposure: quality.exposure,
    sharpness: quality.sharpness,
    explanation: 'Real luminance quality metrics available; native visual-mass candidate is intentionally deferred.'
  };
}

File: modules/frame-analysis/android/src/main/java/com/normacamera/frameanalysis/LumaMetrics.kt
package com.normacamera.frameanalysis

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

internal data class ExposureMetrics(
  val exposureScore: Double,
  val meanLuma: Double,
  val clippedHighlightsRatio: Double,
  val crushedShadowsRatio: Double
)

internal data class SharpnessMetrics(
  val sharpnessScore: Double,
  val edgeEnergy: Double
)

internal data class LumaQualityMetrics(
  val exposure: ExposureMetrics,
  val sharpness: SharpnessMetrics
)

enum class LumaValueRange {
  AUTO,
  UNIT,
  BYTE
}

internal object LumaMetrics {
  private const val TARGET_MEAN_LUMA = 0.52
  private const val MEAN_WEIGHT = 180.0
  private const val HIGHLIGHT_THRESHOLD = 0.96
  private const val SHADOW_THRESHOLD = 0.04
  private const val HIGHLIGHT_PENALTY = 80.0
  private const val SHADOW_PENALTY = 70.0
  private const val SHARPNESS_SCALE = 420.0

  fun compute(
    grid: DoubleArray,
    width: Int,
    height: Int,
    valueRange: LumaValueRange = LumaValueRange.AUTO
  ): LumaQualityMetrics {
    require(width > 0) { "width must be positive" }
    require(height > 0) { "height must be positive" }
    require(grid.size == width * height) { "grid size does not match width and height" }

    val resolvedRange = resolveValueRange(grid, valueRange)
    var sum = 0.0
    var clippedHighlights = 0
    var crushedShadows = 0

    for (value in grid) {
      val luma = toUnitLuma(value, resolvedRange)
      sum += luma
      if (luma >= HIGHLIGHT_THRESHOLD) clippedHighlights += 1
      if (luma <= SHADOW_THRESHOLD) crushedShadows += 1
    }

    val sampleCount = grid.size.toDouble()
    val meanLuma = sum / sampleCount
    val clippedHighlightsRatio = clippedHighlights / sampleCount
    val crushedShadowsRatio = crushedShadows / sampleCount
    val baseExposureScore = 100.0 - abs(meanLuma - TARGET_MEAN_LUMA) * MEAN_WEIGHT
    val exposureScore = clamp(
      baseExposureScore - clippedHighlightsRatio * HIGHLIGHT_PENALTY - crushedShadowsRatio * SHADOW_PENALTY,
      0.0,
      100.0
    )

    var edgeEnergyTotal = 0.0
    var edgeSampleCount = 0

    for (y in 1 until height - 1) {
      for (x in 1 until width - 1) {
        val center = y * width + x
        val left = toUnitLuma(grid[center - 1], resolvedRange)
        val right = toUnitLuma(grid[center + 1], resolvedRange)
        val up = toUnitLuma(grid[center - width], resolvedRange)
        val down = toUnitLuma(grid[center + width], resolvedRange)
        edgeEnergyTotal += abs(right - left) + abs(down - up)
        edgeSampleCount += 1
      }
    }

    val edgeEnergy = if (edgeSampleCount == 0) 0.0 else edgeEnergyTotal / edgeSampleCount
    val sharpnessScore = clamp(edgeEnergy * SHARPNESS_SCALE, 0.0, 100.0)

    return LumaQualityMetrics(
      exposure = ExposureMetrics(
        exposureScore = exposureScore,
        meanLuma = meanLuma,
        clippedHighlightsRatio = clippedHighlightsRatio,
        crushedShadowsRatio = crushedShadowsRatio
      ),
      sharpness = SharpnessMetrics(
        sharpnessScore = sharpnessScore,
        edgeEnergy = edgeEnergy
      )
    )
  }

  private fun resolveValueRange(grid: DoubleArray, valueRange: LumaValueRange): LumaValueRange {
    if (valueRange != LumaValueRange.AUTO) return valueRange
    for (value in grid) {
      if (value.isNaN() || value.isInfinite()) continue
      if (value > 1.0) return LumaValueRange.BYTE
    }
    return LumaValueRange.UNIT
  }

  private fun toUnitLuma(value: Double, valueRange: LumaValueRange): Double {
    if (value.isNaN() || value.isInfinite()) return 0.0
    val normalized = when (valueRange) {
      LumaValueRange.BYTE -> value / 255.0
      LumaValueRange.UNIT -> value
      LumaValueRange.AUTO -> value
    }
    return clamp(normalized, 0.0, 1.0)
  }

  private fun clamp(value: Double, minValue: Double, maxValue: Double): Double {
    return min(maxValue, max(minValue, value))
  }
}
